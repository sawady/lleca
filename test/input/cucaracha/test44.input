
fun factorial(n : Int) : Int {
    i := 0
    p := 1
    while i < n {
        i := i + 1
        p := p * i
    }
    return p
}

fun main() {
  i := 0
  while i <= 10 {
    putNum(factorial(i))
    putChar(10)
    i := i + 1
  }
}

