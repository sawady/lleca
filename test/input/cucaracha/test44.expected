Cons(
  Function(
    factorial,
    Cons(
      Parameter(
        n,
        Int
      ),
      Nil
    ),
    Int,
    Cons(
      StmtAssign(
        i,
        ExprConstNum(
          0
        )
      ),
      Cons(
        StmtAssign(
          p,
          ExprConstNum(
            1
          )
        ),
        Cons(
          StmtWhile(
            ExprLt(
              ExprVar(
                i
              ),
              ExprVar(
                n
              )
            ),
            Cons(
              StmtAssign(
                i,
                ExprAdd(
                  ExprVar(
                    i
                  ),
                  ExprConstNum(
                    1
                  )
                )
              ),
              Cons(
                StmtAssign(
                  p,
                  ExprMul(
                    ExprVar(
                      p
                    ),
                    ExprVar(
                      i
                    )
                  )
                ),
                Nil
              )
            )
          ),
          Cons(
            StmtReturn(
              ExprVar(
                p
              )
            ),
            Nil
          )
        )
      )
    )
  ),
  Cons(
    Function(
      main,
      Nil,
      Unit,
      Cons(
        StmtAssign(
          i,
          ExprConstNum(
            0
          )
        ),
        Cons(
          StmtWhile(
            ExprLe(
              ExprVar(
                i
              ),
              ExprConstNum(
                10
              )
            ),
            Cons(
              StmtCall(
                putNum,
                Cons(
                  ExprCall(
                    factorial,
                    Cons(
                      ExprVar(
                        i
                      ),
                      Nil
                    )
                  ),
                  Nil
                )
              ),
              Cons(
                StmtCall(
                  putChar,
                  Cons(
                    ExprConstNum(
                      10
                    ),
                    Nil
                  )
                ),
                Cons(
                  StmtAssign(
                    i,
                    ExprAdd(
                      ExprVar(
                        i
                      ),
                      ExprConstNum(
                        1
                      )
                    )
                  ),
                  Nil
                )
              )
            )
          ),
          Nil
        )
      )
    ),
    Nil
  )
)
